pkg/ebpf/printk_patcher.go

```diff
@@ -0,0 +1,168 @@
+			for i := idx - 1; i >= maxLookback && stackOffsetIns == nil; i-- {
+				candidate := &p.Instructions[i]
+				if candidate.OpCode == movRegOpCode && candidate.Dst == asm.R1 && candidate.Src == asm.RFP {
+					// Ok, so we found the instruction that loads the stack pointer into r1
+					// From that, advance until we find the add instruction that modifies the stack offset
+					for j := i + 1; j < idx; j++ {
+						candidate = &p.Instructions[j]
+						if candidate.OpCode == addImmOpCode && candidate.Dst == asm.R1 {
+							stackOffsetIns = candidate
+							break
+						}
+					}
+				}
+			}
+			if stackOffsetIns == nil {
+				err = log.Warnf("Could not find stack offset instruction for bpf_trace_printk call %d in %s", idx, p.Name)
+				continue
+			}
+			newlineOffset := int16(stackOffsetIns.Constant + lengthLoadIns.Constant - 1) // -1 because the last character is the null character
+
+			// Now find which store instruction is responsible for putting the newline character
+			stringStoreInsIdx := -1
+			inInstructionOffset := 0
+			for i := idx - 1; i >= maxLookback; i-- {
+				candidate := &p.Instructions[i]
+				if candidate.OpCode.Class() == asm.StXClass && candidate.Src == asm.R1 && candidate.Dst == asm.RFP {
+					if candidate.OpCode.Size() == asm.InvalidSize {
+						err = log.Warnf("BUG: store instruction %v returned asm.InvalidSize", candidate)
+						continue
+					}
+
+					minOffset := candidate.Offset
+					maxOffset := minOffset + int16(candidate.OpCode.Size().Sizeof())
+
+					if newlineOffset >= minOffset && newlineOffset <= maxOffset {
+						// We found the store instruction that loads the newline character, exit the loop
+						stringStoreInsIdx = i
+						inInstructionOffset = int(newlineOffset - minOffset)
+						break
+					}
+				}
+			}
+			if stringStoreInsIdx == -1 {
+				err = log.Warnf("Could not find store instruction for bpf_trace_printk call %d in %s", idx, p.Name)
+				continue
+			}
+
+			// Now try to find the load instruction that loads the string into the R1 register
+			// var stringLoadIns *asm.Instruction
+			for i := stringStoreInsIdx - 1; i >= maxLookback; i-- {
+				candidate := &p.Instructions[i]
+				if (candidate.OpCode == movImmOpCode || candidate.OpCode == ldDWImmOpCode) && candidate.Dst == asm.R1 {
+					// This is the load instruction that's putting the newline character on the stack
+					// (the second to last one, that's why we check <= 9 and not <= 8)
+					mask := uint64(0xff) << (uint64(inInstructionOffset-1) * 8)
+					if candidate.Constant&int64(mask) == int64('\n')<<(uint64(inInstructionOffset-1)*8) {
+						candidate.Constant &= int64(^mask) // Set the newline byte to 0
+
+						// We've correctly patched this instruction, reduce the length in one byte and continue looking for more
+						lengthLoadIns.Constant--
+					}
+					break
+				}
+			}
+		}
+	}
+
+	if err != nil {
+		return fmt.Errorf("at least one error happened patching printk calls: %v", err)
```

Comment:
If you want to collect all the errors, store them in a `[]error` and then use `errors.Join` OR use `errors.Join` every time there is an error.